Soundtrack Events & Rework

PiersSK opened the pull request:
This creates a soundtrack SO and a new specific soundtrack manager that is checking for a set of conditions described by the SO

Please play through the game and check if the following criteria are being met:
- *Opening Soundtrack* should play on start.
- *Ambient Loop* is a few seconds of mysterious sounds. It should play as soon as the opening is done. It should then repeat randomly every 15-45 seconds
- *Phase 2 Begin* should play as soon as phase 2 starts.
- *Phase 2 Loop* should play straight after phase 2 begin
- *Phase 2 Loop Intense* should play after the "unlock weapon" (one with keycards) task is completed, or 80% of the timer has elapsed
- *Final Cadence" should play as soon as you fire, or when 12 seconds are remaining on the clock
- *Pacifist Ending* should instantly pause as soon as the final cadence has finished if you fired, but otherwise it should immediately play

Try to break it - there's some very edge cases where if you SPEED through it might get stuck. Caught a few but needs more consideration I suspect

PiersSK committed:
feat(code): adds soundtrack event SOs

PiersSK committed:
feat(code): creates soundtrack manager for new objects

also creates task and phase completion events for responsive changes

PiersSK committed:
feat(code): adds safety checks for speedy players

checks progression criteria immediately on start to see if anything in the past should have already meant we would expect to skip past this

PiersSK committed:
Merge branch 'develop' into soundtrack-events

PiersSK committed:
fix(code): removes unnecessary editor headers

PiersSK committed:
Merge branch 'develop' into soundtrack-events

JamesYJ95 commented:
Nothing played for me follow Phase 2 Begin.
*Steps to reproduce*
1. Play through P1 relatively slowly, enjoying all the jams
2. Enter P2, confirm Phase 2 Begin plays
3. Do NOT do any of the tasks, wait for Phase 2 Begin to end, so that Phase 2 Loop starts
*Expected*
Phase 2 loop starts
*Observed*
No music plays for the rest of the run. This includes Final Cadence not playing after firing the weapon.

JamesYJ95 committed:
feat(code): Creates Editor folder for custom editors

Required so Unity can ignore these files when compiling, preventing build problems

PiersSK committed:
fix(code): unpauses audio correctly on resume

PiersSK committed:
fix(code): prevents tracks skipping on window defocus

PiersSK committed:
fix(code): adds speed run protection to phase 2 open

PiersSK committed:
fix(code): fixes to pausing background logic

PiersSK committed:
Merge branch 'develop' into soundtrack-events

PiersSK committed:
Merge branch 'develop' into soundtrack-events

PiersSK committed:
fix(code): duplicate scene ids fixed

OrcaOfTime reviewed (approved):
Sounds good
